homeassistant/components/template/sensor.py

```diff
@@ -237,9 +253,23 @@ class TriggerSensorEntity(TriggerEntity, SensorEntity):
     extra_template_keys = (CONF_STATE,)
 
     @property
-    def native_value(self) -> str | None:
+    def native_value(self) -> str | datetime | date | None:
         """Return state of the sensor."""
-        return self._rendered.get(CONF_STATE)
+        state = self._rendered.get(CONF_STATE)
+        if (
+            state is not None
+            and self.device_class
+            in (
+                SensorEntityDeviceClass.DATE,
+                SensorEntityDeviceClass.DATETIME,
+            )
+            and (parsed_state := dt_util.parse_datetime(state)) is not None
```

Comment:
If the parsed datetime is invalid, we should set the sensor to `unavailable` and log a warning.